CIP-0037/CIP-0037.md

```diff
@@ -0,0 +1,352 @@
+Up to date, the most recent, efficient and interesting solutions to provide these are pairing based 
+([Hyperledger Fabric/Idemix standardisation effort](https://hyperledger-fabric.readthedocs.io/en/release-2.2/idemix.html#underlying-cryptographic-protocols), 
+[coconut credentials used by Nym](https://blog.nymtech.net/nyms-coconut-credentials-an-overview-4aa4e922cd51), among 
+others)
+
+
+### Reason for choosing BLS12_381
+The reason for choosing BLS12_381 over BN256 curve is that the former is claimed to provide 128 bits of security, 
+while the latter was reduced to 100 bits of security after the extended number field sieve (a new algorithm to compute 
+the discrete logarithm) was [shown to reduce the security](https://eprint.iacr.org/2016/1102.pdf) of these curves. 
+
+An [EIP](https://eips.ethereum.org/EIPS/eip-2537) for precompiles of curve BLS12_381 already exists, but has been 
+stagnant for a while. Nonetheless, Zcash, MatterLabs and Consensys support BLS12_381 curve, so it is certainly widely 
+used in the space. 
+
+Further reading regarding curve BLS12_381 can be found [here](https://hackmd.io/@benjaminion/bls12-381) and the 
+references thereof cited. For level of detail required in this document, it is sufficient to understand that one can 
+define a pairing over BLS12_381, where a pairing is a map: e:G1 X G2 -> GT, which satisfies the following properties:
+
+* Bilinearity: for all a,b in F^*_q, for all P in G1, Q in G2: e(aP,bQ)=e(P,Q)^(ab)
+* Non-degeneracy: e != 1
+* Computability: There exists an efficient algorithm to compute e
+
+where G1, G2 and GT are three distinct groups of order a prime q. Given that all three groups have the same 
+order, we can refer to the scalars of each group using the same type/structure. 
+
+## Specification
+We now provide the technical specification. 
+
+### Names and types/kinds for the new functions or types
+The added types will be the following, all of which can be represented as a byte array. Even if these types are 
+equivalent to byte arrays of a given size, it makes sense to include these types, to enforce deserialization, and 
+therefore some checks on the data used by the smart contract. In particular, all three types can only be achieved 
+with byte arrays that represent a point which is part of the prime order subgroup. 
+
+* `BLS12_381G1Element`
+* `BLS12_381G2Element`
+* `BLS12_381GTElement`
+
+We need to support the binary operation of G1 and G2 (which are additive groups), as well as the binary operation 
+over GT (which is represented as a multiplicative group). We also want to enable hashing to G1 and G2.
+In particular, we expose the hash to curve (which we denote with `H2G1` and `H2G2` for G1 and G2 
+respectively) algorithm as described in 
+[hash to curve draft](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve#section-8), 
+using `BLS12381G1_XMD:SHA-256_SSWU_RO_` and `BLS12381G2_XMD:SHA-256_SSWU_RO_` for G1 and G2 respectively.
```

Comment:
Shouldn't we use BLAKE2 as is common in Cardano AFAICT? There are a lot of reasons to avoid SHA2, see https://github.com/zcash/zcash/issues/706.

Reply:
We are using the recommendation of the standard for BLS12 381, which is also what is implemented in the blst library. BLS signatures depend on this choice, and the goal is to follow the standard to be compatible with as many implementations of BLS signatures over BLS12 381. 

While Blake is the common and default choice in  Cardano, there are some exceptions. These exceptions are made to follow the standard specifications and make Cardano verification procedures compatible with other implementations. Notable exceptions are ed25519 and VRF (where both use SHA512 to compute the challenge).